doc/inband-telemetry/SAI-Proposal-Advanced-Telemetry-draft.md

```diff
@@ -0,0 +1,740 @@
+
+## Queue reports
+![Queue](figures/queue.png "Figure 5: Queue report")
+
+__Figure 5: Queue Congestion Reports__
+
+Figure 5 shows the queue congestion reports functionality. Switches can be configured to report all packets experience congestion at specified queues. Congestion is detected by queue depth and latency thresholds, which can be configured separately for each individual queue. Queue congestion report is turned off for each queue by default, and is only enabled when depth and latency thresholds are configured.
+
+
+# Specification
+
+This section describes the advanced telemetry API proposal.
+
+## Changes to `saiswitch.h` for switch telemetry attributes
+~~~cpp
+typedef enum _sai_switch_attr_t
+{
+    ......
+    /**
+     * @brief INT endpoint
+     *
+     * @type bool
+     * @flags CREATE_AND_SET
+     * @default False
+     */
+    SAI_SWITCH_ATTR_TELEMETRY_INT_EP_ENABLE,
+
+    /**
+     * @brief INT transit
+     *
+     * @type bool
+     * @flags CREATE_AND_SET
+     * @default False
+     */
+    SAI_SWITCH_ATTR_TELEMETRY_INT_TRANSIT_ENABLE,
+
+    /**
+     * @brief Packet postcard
+     *
+     * @type bool
+     * @flags CREATE_AND_SET
+     * @default False
+     */
+    SAI_SWITCH_ATTR_TELEMETRY_POSTCARD_ENABLE,
+
+    /**
+     * @brief Mirror on Drop
+     *
+     * @type bool
+     * @flags CREATE_AND_SET
+     * @default False
+     */
+    SAI_SWITCH_ATTR_TELEMETRY_MIRROR_ON_DROP_ENABLE,
+
+    /**
+     * @brief Globally unique switch ID
+     *
+     * @type sai_uint32_t
+     * @flags CREATE_AND_SET
+     * @default 0
+     */
+    SAI_SWITCH_ATTR_TELEMETRY_SWITCH_ID,
+
+    /**
+     * @brief Telemetry report source IP address
+     *
+     * @type sai_ip4_t
+     * @flags CREATE_AND_SET
+     */
+    SAI_SWITCH_ATTR_TELEMETRY_REPORT_SRC_IP,
+
+    /**
+     * @brief Telemetry report destination IP addresses
+     *
+     * @type sai_ip4_list_t
+     * @flags CREATE_AND_SET
+     */
+    SAI_SWITCH_ATTR_TELEMETRY_REPORT_DST_IP_LIST,
+
+    /**
+     * @brief Telemetry report virtual router ID
+     *
+     * @type sai_object_id_t
+     * @flags CREATE_AND_SET
+     * @default Switch default virtual router ID
+     */
+    SAI_SWITCH_ATTR_TELEMETRY_REPORT_VIRTUAL_ROUTER_ID,
+
+    /**
+     * @brief Telemetry report truncate size
+     *
+     * @type sai_uint16_t
+     * @flags CREATE_AND_SET
+     * @default 256
+     */
+    SAI_SWITCH_ATTR_TELEMETRY_REPORT_TRUNCATE_SIZE,
+
+    /**
+     * @brief Telemetry report UDP destination port
+     *
+     * @type sai_uint16_t
+     * @flags CREATE_AND_SET
+     * @default: 9999
+     */
+    SAI_SWITCH_ATTR_TELEMETRY_REPORT_UDP_DST_PORT,
+
+    /**
+     * @brief Telemetry flow state clear cycle
+     *
+     * @type sai_uint16_t
+     * @flags CREATE_AND_SET
+     * @default 0
+     */
+    SAI_SWITCH_ATTR_TELEMETRY_FLOW_STATE_CLEAR_CYCLE,
```

Comment:
what is the time granularity here?

Reply:
Currently second in the example. May need to change to millisecond to increase flexibility. Should we specify the granularity in SAI?